openfisca_core/tracers.py

```diff
@@ -1,216 +1,198 @@
+    def __getattr__(self, key):
+        child = getattr(self.parameter_node_at_instant, key)
+        return self.get_traced_child(child, key)
 
-            Value example:
+    def __getitem__(self, key):
+        child = self.parameter_node_at_instant[key]
+        return self.get_traced_child(child, key)
 
-            >>> {'income_tax<2017-01>', 'basic_income<2017-01>'}
+    def get_traced_child(self, child, key):
+        period = self.parameter_node_at_instant._instant_str
+        if isinstance(child, (ParameterNodeAtInstant, VectorialParameterNodeAtInstant)):
+            return TracingParameterNodeAtInstant(child, self.tracer)
+        if not isinstance(key, str) or isinstance(self.parameter_node_at_instant, VectorialParameterNodeAtInstant):
+            # In case of vectorization, we keep the parent node name as, for instance, rate[status].zone1 is best described as the value of "rate"
+            name = self.parameter_node_at_instant._name
+        else:
+            name = '.'.join([self.parameter_node_at_instant._name, key])
+        if isinstance(child, (np.ndarray,) + ALLOWED_PARAM_TYPES):
+            self.tracer.record_parameter_access(name, period, child)
+        return child
 
-        .. py:attribute:: stack
 
-            ``list`` of the calculations that have started, but have not finished. The first item is one of the :attr:`requested_calculations`, and each other item is a dependency of the one preceding him. Note that after a calculation is finished, :attr:`stack` is always ``[]``.
+class SimpleTracer:
 
-            Value example:
+    def __init__(self):
+        self._stack = []
 
-            >>> ['income_tax<2017-01>', 'global_income<2017-01>', 'salary<2017-01>']
+    def enter_calculation(self, variable: str, period):
+        self.stack.append({'name': variable, 'period': period})
 
-        .. py:attribute:: trace
+    def record_calculation_result(self, value: np.ndarray):
+        pass  # ignore calculation result
 
-            ``dict`` containing, for each calculation, its result and its immediate dependencies.
+    def record_parameter_access(self, parameter: str, period, value):
+        pass
 
-            Value example:
+    def exit_calculation(self):
+        self.stack.pop()
 
-            .. code-block:: python
+    @property
+    def stack(self):
+        return self._stack
 
-              {
-                'income_tax<2017-01>': {
-                  'dependencies':['global_income<2017-01>', 'nb_children<2017-01>'],
-                  'parameters' : {'taxes.income_tax_rate<2015-01>': 0.15, ...},
-                  'value': 600
-                  },
-                'global_income<2017-01>': {...}
-              }
 
-        .. py:attribute:: usage_stats
+class FullTracer:
 
-            ``dict`` containing, for each variable computed, the number of times the variable was requested.
+    def __init__(self):
+        self._stacker = SimpleTracer()
```

Comment:
Rename `self._stacker` to `self._simple_tracer` ? I'm not sure what a "stacker" is.

Reply:
The idea here is to say that we use the SimpleTracer for its stack.